File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>

using namespace std;

// 파일명은 100 글자 이내로, 영문 대소문자, 숫자, 공백(" "), 마침표("."), 빼기 부호("-")
vector<string> solution(vector<string> files) {
    vector<string> answer;
    
    // head number tail로 나누어 저장 - 모두 소문자로 변경 
    vector< pair<pair<string, string>, string> > file; 
    for(auto f : files){
        string head = ""; string number = ""; string tail = ""; 
        for(int i = 0; i < f.size(); i++){
            if(isdigit(f[i]) && head == ""){
                head = f.substr(0, i); 
            }
            else if(!isdigit(f[i]) && head != "" && number == ""){
                number = f.substr(head.size(), i-head.size());
                tail = f.substr(i);
                break;
            }
        }
        // tail 부분이 없어서 number 부분에 못 넣었다면
        if(number == "") number = f.substr(head.size()); 
        
        file.push_back({{head, number}, tail});
    }
    
    
    // 정렬(나머지 그대로 유지)
    stable_sort(file.begin(), file.end(), [](const auto &a, const auto &b){
        // head는 모두 소문자로 변경한 뒤 비교
        string ah = ""; string bh = "";
        for(auto h : a.first.first) ah += tolower(h);
        for(auto h : b.first.first) bh += tolower(h);
        if(ah != bh) return ah < bh;
        
        // 앞의 0 제거한 뒤 수만 비교
        int an = stoi(a.first.second); int bn = stoi(b.first.second);
        if(an != bn) return an < bn;
    }); 
    
    // 정답 배열에 넣기
    for(auto f : file){
        answer.push_back(f.first.first+f.first.second+f.second);
    }
    
    return answer;
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*; 

class Body{
    String head; String number; String tail; 
    public Body(String head, String number, String tail){
        this.head = head; this.number = number; this.tail = tail; 
    }
}
class Solution {
    
    // 파일명에 포함된 숫자를 반영한 정렬 기능을 저장소 관리 프로그램에 구현
    public String[] solution(String[] files) {
        
        // HEAD, NUMBER, (TAIL)로 구분
        List<Body> splitedFiles = new ArrayList<>(); 
        for(String f : files){
            String head = "";
            String number = "";
            String tail = ""; 
            for(int i = 0; i < f.length(); i++){
                // head 시작 찾기 (수 시작)
                if(f.charAt(i) >= '0' && f.charAt(i) <= '9'){
                    if(head.equals("")) head = f.substring(0, i); 
                }
                // number 끝 찾기 (수 끝)
                else{
                    if(!head.equals("") && number.equals("")) { // 이미 head가 있음
                        number = f.substring(head.length(), i); 
                        tail = f.substring(i);
                    }
                }
            }
            // number가 아직 안들어왔다면 남은 조각이 number
            if(number.equals("")) number = f.substring(head.length()); 
            // list에 분리한거 삽입
            splitedFiles.add(new Body(head, number, tail)); 
        }
        
        // 정렬
        splitedFiles.sort((a, b) ->{
            String aLowerCaseHead = ""; String bLowerCaseHead = "";
            for(int i = 0; i < a.head.length(); i++) aLowerCaseHead += Character.toLowerCase(a.head.charAt(i));
            for(int i = 0; i < b.head.length(); i++) bLowerCaseHead += Character.toLowerCase(b.head.charAt(i));
            // HEAD 부분을 기준으로 사전 순으로 정렬 (대소문자 구분x)
            if(!aLowerCaseHead.equals(bLowerCaseHead)) return aLowerCaseHead.compareTo(bLowerCaseHead);
            // NUMBER의 숫자 순으로 정렬
            else return Integer.parseInt(a.number) - Integer.parseInt(b.number);
            // HEAD, NUMBER 모두 같을 경우 원래 입력에 주어진 순서를 유지 = stable sort(java는 기본적으로 제공)
        });        
        
        // 다시 합치기 (head + number + tail)
        List<String> answer = new ArrayList<>(); 
        for(Body b : splitedFiles){
            answer.add(b.head + b.number + b.tail);
        }
        
        
        // 정렬된 배열을 출력
        return answer.stream().toArray(String[]::new);
    }
}